DOCSP-30532 Stable API (#25)
mongodb/docs-rust: source/fundamentals/stable-api.txt

.. _rust-stable-api:

==============
{+stable-api+}
==============

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

> .. note::
> 
>    To use {+stable-api+} feature, you must connect to a deployment running MongoDB Server version 5.0 or later.
> 
>    You should only use the {+stable-api+} feature if all the MongoDB
>    servers you are connecting to support this feature.

Reviewer: S: let's change this to an important admonition

```suggestion
.. important::

   To use the {+stable-api+} feature, you must connect to a deployment running MongoDB Server version 5.0 or later.

   You should only use the {+stable-api+} feature if all the MongoDB
   servers you are connecting to support this feature.
```

The {+stable-api+} feature forces the server to run operations with
behaviors that are compatible with the **API version** you specify. An API
version defines the expected behavior of the operations it covers and
the format of server responses. The operations and the server responses
may differ depending on the API version you specify.

When you use the {+stable-api+} feature with an official MongoDB driver, you
can update your driver or server without worrying about backward
compatibility issues of the commands covered by the {+stable-api+}.

To learn more about the commands that the server covers, see
:manual:`{+stable-api+} </reference/stable-api/>` in the Server manual.

Specify an API Version
----------------------

To specify an API version, define a ``ServerApi`` struct and set the ``server_api`` field
of your ``ClientOptions`` instance to this struct.
The ``ServerApi`` struct contains your server API version and options.
To learn more about the options, see the :ref:`Modify Behavior <rust-stable-api-options>` section of this guide.

After you specify an API version, the client runs operations that are compatible with the API version.

.. note::

>    The {+driver-long+} currently only supports {+stable-api+} version 1.

Reviewer: ```suggestion
   The {+driver-long+} currently supports only {+stable-api+} version 1.
```

API Version Specification Example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The following example instantiates a ``Client`` that sets the
> {+stable-api+} version and connects to a server.

Reviewer: S: reword so it doesnt sound like the Client is the thing that is setting the stable api version

Perhaps something like:
```suggestion
The following example sets the {+stable-api+} version when instantiating a ``Client`` 
instance and connects to a server.
```

       |
       | Type: ``bool``
       | Default: ``false``

   * - ``deprecation_errors``
     - | Indicates whether the server should return errors for deprecated features.
       |
       | Type: ``bool``
       | Default: ``false``

.. _rust-stable-api-options-example:

Stable API Options Example
~~~~~~~~~~~~~~~~~~~~~~~~~~

This example enables the {+stable-api+} feature with the following specifications:

- Uses {+stable-api+} version 1
- Returns errors for features that aren't part of version 1
- Returns errors for deprecated features

.. literalinclude:: /includes/fundamentals/code-snippets/stable-api-behavior.rs
   :language: rust
   :dedent:
   :start-after: start-stable-api-behavior
   :end-before: end-stable-api-behavior

Additional Information
----------------------

To learn more about connecting to your MongoDB instance or replica set,
see the :ref:`rust-connect-to-mongodb`.

API Documentation
~~~~~~~~~~~~~~~~~

To learn more about any of the methods or types discussed in this
guide, see the following API Documentation:

- `Client <{+api+}/struct.Client.html>`__
- `ClientOptions <{+api+}/options/struct.ClientOptions.html>`__

> - `ServerAPI() <{+api+}/options/struct.ServerApi.html>`__

Reviewer: S: since this is not a method

```suggestion
- `ServerAPI <{+api+}/options/struct.ServerApi.html>`__
```

> - `ServerApiVersion <{+api+}/options/enum.ServerApiVersion.html>`__

Reviewer: S: also link `with_options()`

.. _rust-stable-api:

==============
{+stable-api+}
==============

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. note::

>    The {+stable-api+} feature requires MongoDB Server 5.0 or later.

Reviewer: ```suggestion
   To use {+stable-api+} feature, you must connect to a deployment running MongoDB Server version 5.0 or later.
```

   You should only use the {+stable-api+} feature if all the MongoDB
   servers you are connecting to support this feature.

Overview
--------

In this guide, you can learn how to specify **{+stable-api+}**
compatibility when connecting to a MongoDB instance or replica set.

> The {+stable-api+} feature forces the server to run operations with
> behaviors compatible with the **API version** you specify. An API

Reviewer: ```suggestion
The {+stable-api+} feature forces the server to run operations with
behaviors that are compatible with the **API version** you specify. An API
```

version defines the expected behavior of the operations it covers and
the format of server responses. The operations and the server responses
may differ depending on the API version you specify.

> When you use the Stable API feature with an official MongoDB driver, you

Reviewer: ```suggestion
When you use the {+stable-api+} feature with an official MongoDB driver, you
```

> When you use the Stable API feature with an official MongoDB driver, you

Reviewer: ```suggestion
When you use the Stable API feature with the {+driver-short+}, you
```

Reviewer: It's only a suggestion, but making sure you saw this because I didn't see any changes in the copy!

can update your driver or server without worrying about backward
compatibility issues of the commands covered by the {+stable-api+}.

> To learn more about the commands the server covers, see
> :manual:`{+stable-api+} </reference/stable-api/>`.

Reviewer: S: we are trying to use the "see xxx in the Server manual" wherever applicable!
```suggestion
To learn more about the commands that the server covers, see
:manual:`{+stable-api+} </reference/stable-api/>` in the Server manual.
```

Specify an API Version
----------------------

> To specify an API version, set the ``server_api`` field of ``mongodb::options::ClientOptions``
> with a ``mongodb::options::ServerApi`` struct.
> The ``ServerApi`` struct contains your :ref:`server API version and options <rust-stable-api-options>`.

Reviewer: S: Only use the name of the type/method. And, make links more clear
```suggestion
To specify an API version, define a ``ServerApi`` struct and set the ``server_api`` field
of your ``ClientOptions`` instance to this struct.
The ``ServerApi`` struct contains your server API version and options.
To learn more about the options, see the :ref:`Modify Behavior <rust-stable-api-options>` section of this guide.
```

> https://docs.rs/mongodb/latest/mongodb/options/struct.ClientOptions.html#structfield.server_api

Reviewer: Did you mean to include this line?

Author: Ah! Nope.

> After you specify an API version, the client runs operations that are compatible with the API version for the duration of your connection.

Reviewer: S: remove redundancy
```suggestion
After you specify an API version, the client runs operations that are compatible with the API version.
```

.. note::

>    The {+driver-long+} currently only supports ``mongodb::options::ServerApiVersion::V1``.

Reviewer: ```suggestion
   The {+driver-long+} currently only supports {+stable-api+} version 1.
```

API Version Specification Example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The following example instantiates a ``Client`` that sets the
{+stable-api+} version and connects to a server.

.. literalinclude:: /includes/fundamentals/code-snippets/stable-api.rs
   :language: rust
   :dedent:
   :start-after: start-stable-api

>    :end-before: end-stable-api

Reviewer: S: constrain this include to start at "let server_api = ..." and end in the line that creates the client

.. _rust-stable-api-options:

Modify Behavior
---------------

> You can further modify the behavior of the {+stable-api+} feature by setting fields on the ``ServerApi`` type.

Reviewer: ```suggestion
You can modify the behavior of the {+stable-api+} feature by setting fields in the ``ServerApi`` struct.
```

> While the fields can be set manually, prefer using the ``ServerApi`` ``builder()`` function.

Reviewer: S: avoid passive voice
```suggestion
While you can set the fields in a ``ServerApi`` struct manually, you can use the builder design pattern to define the struct more efficiently.
```

.. list-table::
   :header-rows: 1
   :stub-columns: 1

>    :widths: 30 70

Reviewer: S: reformat the table here to be like the one used in the [Insert guide](https://github.com/mongodb/docs-rust/blob/c9205ba3b55ea5443af10abfc0129c49cb34c807/source/fundamentals/crud/write-operations/insert.txt#L96-L125)

   * - Field
     - Description

   * - ``version``

>      - | The API version to use.

Reviewer: ```suggestion
     - | The {+stable-api+} version to use.
```

       | Specified with the ``ServerAPIVersion`` enum.
       | 
       | Default: **ServerAPIVersion1**

   * - ``strict``
     - | Indicates whether the server should return errors for features that aren't part of the API version.
       |
       | Default: **false**

   * - ``deprecation_errors``
     - | Indicates whether the server should return errors for deprecated features.
       |
       | Default: **false**

.. _rust-stable-api-options-example:
       
Stable API Options Example
~~~~~~~~~~~~~~~~~~~~~~~~~~

> This example specifies for the server to perform the following actions:

Reviewer: ```suggestion
This example enables the {+stable-api+} feature with the following specifications:
```

> - Use Version 1 of the API
> - Return errors for features that aren't part of Version 1
> - Return errors for deprecated features

Reviewer: ```suggestion
- Uses {+stable-api+} version 1
- Returns errors for features that aren't part of version 1
- Returns errors for deprecated features
```

.. literalinclude:: /includes/fundamentals/code-snippets/stable-api-behavior.rs
   :language: rust
   :dedent:
   :start-after: start-stable-api-behavior

>    :end-before: end-stable-api-behavior

Reviewer: S: constrain this include to start at the options instantiation and end when the client is created. Also, it looks like this code example doesnt include the builder method for `deprecation_errors`

Additional Information
----------------------

To learn more about connecting to your MongoDB instance or replica set,
see :ref:`rust-connect-to-mongodb`.

API Documentation
~~~~~~~~~~~~~~~~~

> For more information on the options in this section, see the following
> API Documentation:

Reviewer: ```suggestion
To learn more about any of the methods or types discussed in this
guide, see the following API Documentation:
```

- `Client <{+api+}/mongo/options#Client>`__
- `ClientOptions <{+api+}/mongo/options#ClientOptions>`__
- `ServerAPI() <{+api+}/mongo/options#ServerAPI>`__
- `ServerAPIOptions <{+api+}/mongo/options#ServerAPIOptions>`__
- `ServerApiVersion <{+api+}/mongo/options#ServerAPIVersion>`__
- `SetDeprecationErrors() <{+api+}/mongo/options#ServerAPIOptions.SetDeprecationErrors>`__

> - `SetStrict() <{+api+}/mongo/options#ServerAPIOptions.SetStrict>`__

Reviewer: I: these links are all broken
S: update the links to use the Rust API documentation

To learn more about the options, see the :ref:`Modify Behavior <rust-stable-api-options>` section of this guide.

After you specify an API version, the client runs operations that are compatible with the API version.

.. note::

   The {+driver-long+} currently supports only {+stable-api+} version 1.

API Version Specification Example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The following example sets the {+stable-api+} version when instantiating a ``Client``
and connects to a server.

.. literalinclude:: /includes/fundamentals/code-snippets/stable-api.rs
   :language: rust
   :dedent:
   :start-after: start-stable-api
   :end-before: end-stable-api

.. _rust-stable-api-options:

Modify Behavior
---------------

You can modify the behavior of the {+stable-api+} feature by setting fields in the ``ServerApi`` struct.

While you can set the fields in a ``ServerApi`` struct manually, you can use the builder design pattern to define the struct more efficiently.

.. list-table::
   :header-rows: 1
   :stub-columns: 1
   :class: compatibility-large

   * - Field
     - Description

   * - ``version``
     - | The {+stable-api+} version to use.

>        | Specified with the ``ServerAPIVersion`` enum.

Reviewer 2: Suggest replacing `ServerAPIVersion` with `ServerApiVersion` to match Rust driver API casing.

>        | Type: ``ServerAPIVersion``

Reviewer 2: Suggest replacing `ServerAPIVersion` with `ServerApiVersion` to match Rust driver API casing.

>        | Default: ``ServerAPIVersion1``

Reviewer 2: `version` is required. Suggest removing:

> Default: ``ServerAPIVersion1``

       |
       | Type: ``bool``
       | Default: ``false``

   * - ``deprecation_errors``
     - | Indicates whether the server should return errors for deprecated features.
       |
       | Type: ``bool``
       | Default: ``false``

.. _rust-stable-api-options-example:

Stable API Options Example
~~~~~~~~~~~~~~~~~~~~~~~~~~

This example enables the {+stable-api+} feature with the following specifications:

- Uses {+stable-api+} version 1
- Returns errors for features that aren't part of version 1
- Returns errors for deprecated features

.. literalinclude:: /includes/fundamentals/code-snippets/stable-api-behavior.rs
   :language: rust
   :dedent:
   :start-after: start-stable-api-behavior
   :end-before: end-stable-api-behavior

Additional Information
----------------------

To learn more about connecting to your MongoDB instance or replica set,
see the :ref:`rust-connect-to-mongodb`.

API Documentation
~~~~~~~~~~~~~~~~~

To learn more about any of the methods or types discussed in this
guide, see the following API Documentation:

- `Client <{+api+}/struct.Client.html>`__
- `ClientOptions <{+api+}/options/struct.ClientOptions.html>`__

> - `ServerAPI <{+api+}/options/struct.ServerApi.html>`__

Reviewer 2: Suggest replacing `ServerAPI` with `ServerApi` to match Rust driver API casing.
